torchbiggraph/storage_repository.py

```diff
@@ -0,0 +1,255 @@
+            if 'r' in mode:
+                if Path(filepath).exists() and not h5py.is_hdf5(filepath):
+                    raise ValueError('Invalid .h5 file', filepath)
+            return h5py.File(filepath, mode)
+        else:
+            return open(filepath, mode)
+
+
+class HDFSFileContextManager(AbstractContextManager):
+    def __init__(self, path: Path, mode: str, **kwargs) -> None :
+        self._path: Path = path
+        self.mode = mode
+        self.kwargs = kwargs
+
+    def __enter__(self) -> Union[h5py.File, TextIOWrapper]:
+        reload = True
+        if Constants.RELOAD in self.kwargs:
+            reload = self.kwargs[Constants.RELOAD]
+
+        if Constants.READ_MODE in self.mode:
+            self.get_from_hdfs(reload)
+
+        self._file = LocalFileContextManager.get_resource(str(self._path), self.mode)
+        return self._file
+
+
+    def __exit__(self, exception_type: Optional[Type[BaseException]],
+                 exception_value: Optional[BaseException],
+                 traceback: Optional[TracebackType]) -> bool:
+        self._file.close()
+        self._file = None
+
+        if any(ext in self.mode for ext in Constants.WRITE_MODES):
+            self.put_to_hdfs()
+
+        return True
+
+    @staticmethod
+    def get_hdfs_path(path) -> str:
+        return str(path).replace(str(Path.cwd()), '', 1)
+
+    @staticmethod
+    def hdfs_file_exists(path: str) -> bool:
+        rcode = 1
+        try:
+            rcode, output, errors = run_external_cmd(Constants.TEST_FILE.format(remote_path=path))
+        except:
+            pass
+
+        return rcode == 0
+
+    def get_from_hdfs(self, reload):
+        local_path = self._path
+        hdfs_loc = self.get_hdfs_path(local_path)
+
+        if hdfs_loc != str(local_path):
+            # check if hdfs file exists before running get command
+            if self.hdfs_file_exists(hdfs_loc):
+                local_path.parent.mkdir(parents=True, exist_ok=True)
+
+                if local_path.exists():
+                    if reload:
+                        local_path.unlink()
+                        run_external_cmd(Constants.GET.format(remote_path=hdfs_loc, local_path=str(local_path)))
+                    else:
+                        logger.info(f"Skip get: reload is {reload} and local file exists : {local_path}")
+                else:
+                    run_external_cmd(Constants.GET.format(remote_path=hdfs_loc, local_path=str(local_path)))
+            else:
+                raise FileNotFoundError(f"File {hdfs_loc} not found.")
+        else:
+            logger.info(f"identical local and hdfs path. Skipping get {str(local_path)}")
+
+    def put_to_hdfs(self):
+        local_loc_str = str(self._path)
+        hdfs_loc = self.get_hdfs_path(local_loc_str)
+
+        if self._path.exists():
+            if hdfs_loc != local_loc_str:
+                run_external_cmd(Constants.PUT.format(local_path=local_loc_str, remote_path=hdfs_loc))
```

Comment:
Is it really the case that the only way to write to HDFS is to call hadoop in a subprocess on a local file? There's no way to directly stream to/from HDFS?